### DISCUSSION TITLE
[HikariCP] How to set resource.type of metrics as cloud_run_revision instead of generic_node

### DISCUSSION THREAD

**Discussion by KengoTODA** [open-telemetry/opentelemetry-java-instrumentation#10035] [Q&A]
Hello 👋 

I'm using this library to export metrics of DB connection pool on Google Cloud.
I found the following error in otel-collector processes (Cloud Run instances) and want to fix this problem:

> One or more TimeSeries could not be written: Points must be written in order.
> One or more of the points specified had an older start time than the most recent point.:
> generic_node{node_id:localhost,location:global,namespace:} timeSeries[0-153]: 
> workload.googleapis.com/db.client.connections.use_time{pool_name:my-database,instrumentation_version:1.32.0-alpha,instrumentation_source:io.opentelemetry.hikaricp-3.0,service_name:my-service}
> errors:{status:{code:3}  point_count:13}

According to https://stackoverflow.com/a/61108971, it means that otel-collector received DB connection pool metrics from two or more instances, and timestamp of metrics from them are mixed so they seems invalid for the collector.

So I want to try to change resource type of these metrics from `generic_node` to `cloud_run_revision`, or add the instance ID to attribute (or label?) of metrics.

I expected that setting `Attributes` would realize it, however, I found no public API handling `Attributes` in `io.opentelemetry.instrumentation.hikaricp.v3_0` package.

I've also checked https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/6209#issuecomment-1167119142 but I think `Baggage` API is not suitable because DB connection pool metrics are not related to trace/span.

Please help me to find solutions to export HikariCP metrics from multiple instances.
Thanks in advance!

**Comment by laurit** [comment]
Usually one would configure different `service.name` or some other resource attribute as described in https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md#opentelemetry-resource to tell services apart.

**Reply by KengoTODA** [reply]
Thanks for your comment! In this case, every cloud run instances belong to the same service so I think `service.name` resource will not be help. And actually the `service_name` is already included in the metrics:

> workload.googleapis.com/db.client.connections.use_time{pool_name:my-database,instrumentation_version:1.32.0-alpha,instrumentation_source:io.opentelemetry.hikaricp-3.0,**service_name:my-service**}


I will try to add instance ID or other attribute that is unique among each cloud run instance.

**Reply by KengoTODA** [reply]
Seems that Google Cloud official resource provider provides no support for Cloud Run:
https://github.com/GoogleCloudPlatform/opentelemetry-operations-java/blob/fed23bf5411a8420024633f57266b8ab616d8fee/detectors/resources/README.md?plain=1#L7-L20

**Reply by KengoTODA** [reply]
According to https://github.com/GoogleCloudPlatform/opentelemetry-operations-java/pull/250 the  `generic_node` is used intensionally.
I've shared my case in their project: https://github.com/GoogleCloudPlatform/opentelemetry-operations-java/issues/247#issuecomment-1849478718
